File: branches/operator/src/herdstat.cc
/*
 * herdstat -- src/herdstat.cc
 * $Id$
 * Copyright (c) 2005 Aaron Walker <ka0ttic at gentoo.org>
 *
 * This file is part of herdstat.
 *
 * herdstat is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * herdstat is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * herdstat; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 325, Boston, MA  02111-1257  USA
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <iostream>
#include <map>
#include <vector>
#include <algorithm>
#include <memory>
#include <cstring>
#include <cstdio>
#include <cstdlib>
#include <cstdarg>
#include <cerrno>
#include <ctime>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#ifdef HAVE_GETOPT_H
# include <getopt.h>
#endif

#include "herds.hh"
#include "formatter.hh"
#include "xmlparser.hh"
#include "options.hh"
#include "util.hh"
#include "exceptions.hh"
#include "herds_xml_handler.hh"
#include "action_herd_handler.hh"
#include "action_pkg_handler.hh"
#include "action_dev_handler.hh"
#include "action_stats_handler.hh"

static const std::string default_herdsxml =
    "http://www.gentoo.org/cgi-bin/viewcvs.cgi/misc/herds.xml?rev=HEAD;cvsroot=gentoo;content-type=text/plain";

static const char *short_opts = "H:o:hVvDdtpqfc";

#ifdef HAVE_GETOPT_LONG
static struct option long_opts[] =
{
    {"version",	    no_argument,	0,  'V'},
    {"help",	    no_argument,	0,  'h'},
    {"verbose",	    no_argument,	0,  'v'},
    {"quiet",	    no_argument,	0,  'q'},
    {"debug",	    no_argument,	0,  'D'},
    {"count",	    no_argument,	0,  'c'},
    /* force a fetch of herds.xml */
    {"fetch",	    no_argument,	0,  'f'},
    /* time how long it takes for XML parsing */
    {"timer",	    no_argument,	0,  't'},
    /* instead of displaying devs for a herd, display herds for a dev */
    {"dev",	    no_argument,	0,  'd'},
    /* specify the location of a local herds.xml */
    {"herdsxml",    required_argument,	0,  'H'},
    /* show package stats for the specified herds */
    {"package",	    no_argument,	0,  'p'},
    /* specify a file to write the output to */
    {"outfile",	    required_argument,	0,  'o'},
    { 0, 0, 0, 0 }
};
#endif /* HAVE_GETOPT_LONG */

void
version()
{
    std::cout << PACKAGE << "-" << VERSION << std::endl;
}

void
usage()
{
    std::cerr
	<< "usage: " << PACKAGE << " [options] [args]" << std::endl
	<< "Use --help to see more detailed usage information." << std::endl;
}

void
help()
{
    std::cout
	<< "usage: " << PACKAGE << " [options] [args]" << std::endl

#ifdef HAVE_GETOPT_LONG
	<< " -h, --help            Display this help message." << std::endl
	<< " -V, --version         Display version information." << std::endl
	<< std::endl
	<< "Where [options] can be any of the following:" << std::endl
	<< " -p, --package         Look up packages by herd." << std::endl
	<< " -d, --dev             Look up herds by developer." << std::endl
	<< " -H, --herdsxml <file> Specify location of herds.xml." << std::endl
	<< " -o, --outfile  <file> Send output to the specified file" << std::endl
	<< "                       instead of stdout." << std::endl
	<< " -f, --fetch           Force a fetch of herds.xml." << std::endl
	<< " -v, --verbose         Display verbose output." << std::endl
	<< " -q, --quiet           Don't display labels and fancy colors. Use this"
	<< std::endl
	<< "                       option to pipe herdstat output to other programs"
	<< std::endl
	<< " -D, --debug           Display debugging messages." << std::endl
	<< " -t, --timer           Display elapsed time of XML parsing." << std::endl
	<< " -c, --count           Instead of displaying data items, display the" << std::endl
	<< "                       number of items." << std::endl
	<< " -c, --count           Display the number of items instead of the" << std::endl
	<< "                       items themself." << std::endl
	<< std::endl
	<< "Where [args] depends on the specified action:" << std::endl
	<< " default action        1 or more herds." << std::endl
	<< " -p, --package         1 or more herds." << std::endl
	<< " -d, --dev             1 or more developers." << std::endl
	<< "Both the default action and the --dev action support an 'all' target" << std::endl
	<< "that show all of the devs or herds.  If both --dev and --package are" << std::endl
	<< "specified, " << PACKAGE << " will display all packages maintained by" << std::endl
	<< "the specified developer." << std::endl
#else

	<< " -h              Display this help message." << std::endl
	<< " -V              Display version information." << std::endl
	<< std::endl
	<< "Where [options] can be any of the following:" << std::endl
	<< " -p              Look up packages by herd." << std::endl
	<< " -d              Look up herds by developer." << std::endl
	<< " -H <file>       Specify location of herds.xml." << std::endl
	<< " -o <file>       Send output to the specified file" << std::endl
	<< "                 instead of stdout." << std::endl
	<< " -f              Force a fetch of herds.xml." << std::endl
	<< " -v              Display verbose output." << std::endl
	<< " -q              Don't display labels and fancy colors. Use this"
	<< std::endl
	<< "                 option to pipe herdstat output to other programs"
	<< std::endl
	<< " -D              Display debugging messages." << std::endl
	<< " -t              Display elapsed time of XML parsing." << std::endl
	<< " -c              Display the number of items instead of the" << std::endl
	<< "                 items themself." << std::endl
	<< std::endl
	<< "Where [args] depends on the specified action:" << std::endl
	<< " default action  1 or more herds." << std::endl
	<< " -p              1 or more herds." << std::endl
	<< " -d              1 or more developers." << std::endl
	<< "Both the default action and the -d action support an 'all' target" << std::endl
	<< "that show all of the devs or herds.  If both -d and -p are specified," << std::endl
	<< PACKAGE << " will display all packages maintained by the specified" << std::endl
	<< "developer." << std::endl

#endif /* HAVE_GETOPT_LONG */

	<< std::endl
	<< PACKAGE << " respects the HERDS environment variable." << std::endl
	<< "Set it in your shell rcfile to permanently set the location of "
	<< "your herds.xml" << std::endl;
}

int
handle_opts(int argc, char **argv, std::vector<std::string> *args)
{
    int key, opt_index = 0;

    while (true)
    {
#ifdef HAVE_GETOPT_LONG
	key = getopt_long(argc, argv, short_opts, long_opts, &opt_index);
#else
	key = getopt(argc, argv, short_opts);
#endif /* HAVE_GETOPT_LON)G */

	if (key == -1)
	    break;

	switch (key)
	{
	    /* --dev */
	    case 'd':
		if (optget("action", options_action_T) != action_unspecified and
		    optget("action", options_action_T) != action_pkg)
		    throw args_one_action_only_E();
		if (optget("action", options_action_T) == action_pkg)
		    optset("dev", bool, true);
		else
		    optset("action", options_action_T, action_dev);
		break;
	    /* --package */
	    case 'p':
		if (optget("action", options_action_T) != action_unspecified and
		    optget("action", options_action_T) != action_dev)
		    throw args_one_action_only_E();
		if (optget("action", options_action_T) == action_dev)
		    optset("dev", bool, true);
		optset("action", options_action_T, action_pkg);
		break;
	    /* --outfile */
	    case 'o':
		if (strcmp(optarg, "stdout") != 0)
		{
		    if (strcmp(optarg, "stderr") == 0)
			optset("outstream", std::ostream *, &std::cerr);
		    optset("outfile", std::string, optarg);
		    optset("quiet", bool, true);
		    optset("timer", bool, false);
		}
		break;
	    /* --fetch */
	    case 'f':
		optset("fetch", bool, true);
		break;
	    /* --verbose */
	    case 'v':
		optset("verbose", bool, true);
		break;
	    /* --quiet */
	    case 'q':
		optset("quiet", bool, true);
		break;
	    /* --count */
	    case 'c':
		optset("count", bool, true);
		optset("quiet", bool, true);
		break;
	    /* --herdsxml */
	    case 'H':
		optset("herds.xml", std::string, optarg);
		break;
	    /* --debug */
	    case 'D':
		optset("timer", bool, true);
		optset("debug", bool, true);
		break;
	    /* --timer */
	    case 't':
		if (optget("outfile", std::string) == "stdout")
		    optset("timer", bool, true);
		break;
	    /* --version */
	    case 'V':
		throw args_version_E();
		break;
	    /* --help */
	    case 'h':
		throw args_help_E();
		break;
	    case 0:
		throw args_usage_E();
		break;
	    default:
		throw args_E();
		break;
	}
    }

    if (optind < argc)
    {
	while (optind < argc)
	    args->push_back(argv[optind++]);
    }
    else if (optget("action", options_action_T) != action_unspecified)
	throw args_usage_E();

    return 0;
}

/* unary predicate for the remove_if() call below */
bool isNotAll(std::string &s) { return (s != "all"); }

int
main(int argc, char **argv)
{
    options_T options;
    util::timer_T timer;
    std::string fetched_location = util::sprintf("%s/%s/herds.xml",
	LOCALSTATEDIR, PACKAGE);

    /* try to determine current columns, otherwise use default */
    optset("maxcol", size_t, util::getcols());

    /* HERDS */
    char *result = getenv("HERDS");
    if (result)
	optset("herds.xml", std::string, result);
    else
    {
	/* if a fetched copy exists and is newer than 24hrs, use it.
	 * see the explanation below where we fetch as to why
	 * we go through the trouble of doing this. */
	struct stat s;
	if ((stat(fetched_location.c_str(), &s) == 0) and
	    ((time(NULL) - s.st_mtime) < 86400) and (s.st_size > 0))
	    optset("herds.xml", std::string, fetched_location);
    }

    try
    { 
	std::vector<std::string> nonopt_args;

	/* handle command line options */
	if (handle_opts(argc, argv, &nonopt_args) != 0)
	    throw args_E();

	if (nonopt_args.empty())
	{
	    optset("action", options_action_T, action_stats);
	    optset("quiet", bool, false);
	}

	/* remove duplicates; also has the nice side advantage
	 * of sorting the output				*/
	std::sort(nonopt_args.begin(), nonopt_args.end());
	std::vector<std::string>::iterator pos =
	    std::unique(nonopt_args.begin(), nonopt_args.end());
	if (pos != nonopt_args.end())
	    nonopt_args.erase(pos);

	/* did the user specify the all target? */
	pos = std::find(nonopt_args.begin(), nonopt_args.end(), "all");
	if (pos != nonopt_args.end())
	{
	    /* yep, so remove everything but it */
	    pos = std::remove_if(nonopt_args.begin(), nonopt_args.end(),
		isNotAll);
	    if (pos != nonopt_args.end())
		nonopt_args.erase(pos);

	    optset("all", bool, true);
	}

	if (optget("debug", bool))
	    options.dump(*optget("outstream", std::ostream *));

	if (not optget("quiet", bool) and optget("fetch", bool) and nonopt_args.empty())
	    optset("verbose", bool, true);

	/* every action handler needs to parse herds.xml for one reason
	 * or another, so let's get it over with. */
	struct stat s;
	try
	{
	    if (optget("fetch", bool) and
		(optget("herds.xml", std::string).find("://") == std::string::npos))
		optset("herds.xml", std::string, default_herdsxml);

	    if (optget("herds.xml", std::string).find("://") != std::string::npos)
	    {
		/* NOTE: ideally we'd love to only fetch it when the timestamp
		 * and size are different, however, because the default
		 * location to download herds.xml is from ViewCVS, there is no
		 * Last-Modified header, meaning wget can't do timestamps :( */

		if (not optget("quiet", bool))
		{
		    std::cout
			<< "Fetching herds.xml..."
			<< std::endl << std::endl;
		}

		/* backup cached copy if it exists so we can use it
		 * if fetching fails */
		if (util::is_file(fetched_location))
		    util::copy_file(fetched_location, fetched_location + ".bak");

		/* fetch it */
		if (util::fetch(optget("herds.xml", std::string), fetched_location,
		    optget("verbose", bool)) != 0)
		    throw fetch_E();

		/* because we tell wget to clobber the file, if fetching fails for
		 * some reason, it'll truncate the old one - make sure the file
		 * is greater than 0 bytes */
		if ((stat(fetched_location.c_str(), &s) == 0) and (s.st_size > 0))
		    optset("herds.xml", std::string, fetched_location);
		else
		    throw fetch_E();

		/* remove back up copy */
		unlink((fetched_location + ".bak").c_str());
	    }
	}
	catch (const fetch_E &e)
	{
	    std::cerr << "Error fetching " << optget("herds.xml", std::string)
		<< std::endl << std::endl;

	    /* if we can't fetch it but have an old cached copy, use it */
	    if (util::is_file(fetched_location + ".bak"))
	    {
		std::cerr << "Using cached copy... ";
		util::move_file(fetched_location + ".bak", fetched_location);
		optset("herds.xml", std::string, fetched_location);
	    }
		
	    std::cerr
		<< "If you have a local copy, specify it using -H or by "
		<< std::endl << "setting the HERDS environment variable."
		<< std::endl;

	    if (stat(fetched_location.c_str(), &s) != 0)
		return EXIT_FAILURE;
	    else if (s.st_size == 0)
	    {
		unlink(fetched_location.c_str());
		return EXIT_FAILURE;
	    }
	    else
		std::cerr << std::endl;
	}

	/* make sure herds.xml exists */
	if (not util::is_file(optget("herds.xml", std::string)))
	    throw bad_fileobject_E("%s: %s", optget("herds.xml", std::string).c_str(),
		strerror(errno));

	/* if there's no nonoption args, we're done */
	if (optget("fetch", bool) and nonopt_args.empty())
	    return EXIT_SUCCESS;

	std::auto_ptr<HerdsXMLHandler_T> handler(new HerdsXMLHandler_T());
	try
	{
	    if (optget("timer", bool))
		timer.start();

	    /* parse herds.xml */
	    XMLParser_T parser(&(*handler));
	    parser.parse(optget("herds.xml", std::string));

	    if (optget("timer", bool))
		timer.stop();
	}
	catch (const XMLParser_E &e)
	{
	    std::cerr << "Error parsing '" << e.file() << "': "	<< e.error()
		<< std::endl;
	    return EXIT_FAILURE;
	}

	/* setup outfile */
	std::ostream *outstream = NULL;
	if (optget("outfile", std::string) != "stdout" and
	    optget("outfile", std::string) != "stderr")
	{
	    outstream = new std::ofstream(optget("outfile", std::string).c_str());
	    if (not *outstream)
		throw bad_fileobject_E("%s: %s", optget("outfile", std::string).c_str(),
		    strerror(errno));
	    optset("outstream", std::ostream *, outstream);
	}

	/* set common format attributes */
	util::color_map_T color;
	formatter_T output(outstream);
	output.set_colors(true);
	output.set_quiet(optget("quiet", bool));
	output.set_labelcolor(color[green]);

	/* set default action */
	if (optget("action", options_action_T) == action_unspecified)
	    optset("action", options_action_T, action_herd);

	std::map<options_action_T, action_handler_T * > handlers;
	handlers[action_herd]  = new action_herd_handler_T();
	handlers[action_dev]   = new action_dev_handler_T();
	handlers[action_pkg]   = new action_pkg_handler_T();
	handlers[action_stats] = new action_stats_handler_T();

	action_handler_T *action_handler =
	    handlers[optget("action", options_action_T)];
	if (action_handler)
	{
	    try
	    {
		if ((*action_handler)(handler->herds, nonopt_args) != EXIT_SUCCESS)
		    return EXIT_FAILURE;
	    }
	    catch (action_E)
	    {
		return EXIT_FAILURE;
	    }
	}
	else
	    throw args_unimplemented_E();

	if (outstream)
	    delete outstream;

	if (optget("timer", bool))
	    throw timer_E();
    }
    catch (const bad_fileobject_E &e)
    {
	std::cerr << e.what() << std::endl;
	return EXIT_FAILURE;
    }
    catch (const timer_E)
    {
	std::cout
	    << "Took " << timer.elapsed() << "ms to parse herds.xml." << std::endl;
    }
    catch (const args_help_E)
    {
	help();
	return EXIT_SUCCESS;
    }
    catch (const args_version_E)
    {
	version();
	return EXIT_SUCCESS;
    }
    catch (const args_usage_E)
    {
	usage();
	return EXIT_FAILURE;
    }
    catch (const args_E)
    {
	usage();
	return EXIT_FAILURE;
    }
    catch (const herdstat_base_E &e)
    {
	std::cerr << "Unhandled exception: " << e.what() << std::endl;
	return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

File: tags/release-0.1/pym/herdstat.py
#!/usr/bin/python -O
# Copyright 1999-2005 Gentoo Foundation
# Distributed under the terms of the GNU General Public License v2
# $Header: $

# $Id: herdstat.py 66 2005-03-02 15:44:57Z ka0ttic $

# display statistics about given herd
# Author: Aaron Walker <ka0ttic@gentoo.org>

__author__      = "Aaron Walker"
__email__       = "ka0ttic@gentoo.org"
__productname__ = "herdstat"
__description__ = "display herd statistics"
__version__     = "0.1"

import os
import sys
from xml.sax import make_parser,SAXException

sys.path.insert(0, "/usr/lib/portage/pym")
import portage
from output import *
from portage_herds import Herds_XML
#from portage_metadata import Metadata_XML

settings = portage.config(clone=portage.settings)

herds_url = "http://www.gentoo.org/cgi-bin/viewcvs.cgi/misc/herds.xml?rev=HEAD;cvsroot=gentoo;content-type=text/plain"

def print_usage():
    print "usage: " + __productname__ + " [options] herd(s)\n"
    print " -v,--verbose    Display verbose output"
    print " -p,--package    Show statistics regarding packages for specified herd."
    print "                 Note that this option may take a while."
    print " -h,--help       Display this help message"
    print '\n' + __productname__ + " will check the HERDS environment variable for the"
    print "location of a local herds.xml.  If it is not set, or the file does"
    print "not exist, " + __productname__ + " will try to retrieve it from gentoo.org"

def get_herds_xml(filename):
    """open herds.xml and return file object"""

    # TODO: hook into the new portage transport stuff (pym/transports)

    try:
	if filename[:4] == "http":
	    from urllib import urlopen
	    f = urlopen(filename)
	else:
	    f = open(filename)
    except IOError:
	print red("!!! Failed to open '%s'." % filename)
	return None

    return f

def parse_xml(handler, xml):
    """parse given file object with the given content handler and
       return finished handler"""

    parser = make_parser()
    parser.setContentHandler(handler)

    try:
	parser.parse(xml)
    except SAXException, e:
	print red("!!! Error parsing xml: " + e.getMessage())
	return None

    return handler

def format(prefix, string):
    """Given a prefix and a string, return a string with the appropriate
       amount of whitespace in between"""

    # TODO: use formatters/writers so we don't have to do this crap

    tmpstr = green(prefix)
    if tmpstr != "":
	tmpstr += ":"

    tmpstr = tmpstr.ljust(32) + string
    return tmpstr

def display_dev_data(herd, handler):
    """Display all devs that work on the specified herd."""

    # add some padding for the extra chars present from the coloring
    maxcol = 78 + 20
    if os.environ.has_key('COLUMNS'):
	maxcol = os.environ['COLUMNS'] - 2 + 20

    ndev = len(handler._herds[herd])
    outstr = green("Members(%d)" % ndev)
    if ndev < 10:
	outstr += ":".ljust(3)
    else:
	outstr += ":".ljust(2)
	
    devs = handler._herds[herd].keys()
    devs.sort()

    for dev in devs:
	if verbose:
	    outstr += '\n' + "".ljust(13) + dev + '\n'
	    outstr += "".ljust(13) + handler._herds[herd][dev][0] + '\n'
	    if len(handler._herds[herd][dev]) > 1:
		outstr += "".ljust(13) + handler._herds[herd][dev][1] + '\n'
	else:
	    dev = dev.split('@')[0]

	    # only append if there's enough room
	    if len(outstr) + len(dev) < maxcol:
		outstr += dev + " "
	    # otherwise, it's full; print it, and start a new line with the dev
	    else:
		print outstr
		outstr = "".ljust(13) + dev + " "

    # empty the buffer
    if len(outstr) != 0:
	print outstr

def display_pkg_data(herd, handler):
    """Display all packages that belong in the specified herd."""
    from glob import glob

    npkg = 0
    packages = []
    for category in glob(settings['PORTDIR']+'/*-*'):
	# this isn't the prettiest, but it's the fastest
	out = os.popen('grep -l \<herd\>'+herd+'\</herd\> '+category+'/*/metadata.xml')
	for metadata in out.readlines():
	    npkg += 1
	    p = metadata.split('/')
	    packages.append(p[-3] + '/' + p[-2])
    
    print green("Packages(%d)" % npkg) + ":"

    packages.sort()
    for pkg in packages:
	print "".ljust(13) + pkg

def display_herd_data(herd, handler):
    """Display herd statistics for the specified herd"""

    # does herd exist?
    if not handler._herds.has_key(herd):
	print "Herd '%s' was not found in herds.xml." % herd
	return 1

    print format("Herd", herd)
    print format("Description", handler._descs[herd])
    
    if package:
	display_pkg_data(herd, handler)
    else:
	display_dev_data(herd, handler)

    return 0

def herdstat(herds):
    """display statistics about given herd(s)/dev(s)"""

    # allow user to use local file over URL if we can
    # or even an alternate URL
    if os.environ.has_key('HERDS'):
	herds_xml = get_herds_xml(os.environ['HERDS'])
    else:
	print "HERDS environment variable is not set."
	print "Attempting to retrieve herds.xml from http://www.gentoo.org/."
	herds_xml = get_herds_xml(herds_url)

    if herds_xml == None:
	return 1

    handler = parse_xml(Herds_XML(), herds_xml)
    if handler == None:
	return 1

    herds_xml.close()

    for herd in herds:
	if display_herd_data(herd, handler) != 0:
	    return 1
	# only print seperator newlines if we're not displaying the last one
	if herd != herds[-1]:
	    print '\n'

    return 0

if __name__ == '__main__':
    from getopt import gnu_getopt,GetoptError

    if len(sys.argv) == 1:
	print_usage()
	sys.exit(1)

    try:
	opts,args = gnu_getopt(sys.argv[1:], "hvp", ["help","verbose","package"])
    except GetoptError:
	print_usage()
	sys.exit(1)

    verbose = False
    package = False

    for o, a in opts:
	if o in ("-v", "--verbose"):
	    verbose = True
	if o in ("-h", "--help"):
	    print_usage()
	    sys.exit(0)
	if o in ("-p", "--package"):
	    package = True

    # all non-option args are interpreted as herds
    sys.exit(herdstat(args))
